torch/distributed/_sharding_spec/api.py

```diff
@@ -96,3 +96,173 @@ def placement(self) -> ShardPlacement:
         Retrieves the shard placement.
         """
         return self._placement
+
+class GenericShardingSpec(PlacementSpec):
+
+    class Shard:
+        """
+        Represents a shard of the overall Tensor including its
+        offsets, lengths and device placement.
+
+        Args:
+            shard_offsets(List[int]): Offsets in the orignal tensor indicating
+                the start offsets for this shard. Should have the same rank as
+                the original tensor.
+            shard_lengths(List[int]): Lengths indicating the length of each
```

Comment:
curious, do we have any requirements/guideline on length distribution across shards? I assume we encourage uniform distribution, but support arbitrary distribution?

Reply:
There are no guidelines as such as part of the specification itself. I think when we build more higher level APIs on top of this, we could have guidelines at that level in terms of what is best for performance.